packages/xsnap/src/xsnap.js

```diff
@@ -96,7 +102,65 @@ export async function xsnap(options) {
   }
 
   /** @type {(opts: import('tmp').TmpNameOptions) => Promise<string>} */
-  const ptmpName = promisify(fs.tmpName);
+  const ptmpName = fs.tmpName && promisify(fs.tmpName);
+
+  const makeLoadSnapshotHandlerWithFS = async () => {
+    assert(snapshotStream);
+    const snapPath = await ptmpName({
+      template: `load-snapshot-${safeHintFromDescription(
+        snapshotDescription,
+      )}-XXXXXX.xss`,
+    });
+
+    const cleanup = async () => fs.unlink(snapPath);
+    const afterSpawn = async () => {};
+
+    try {
+      // eslint-disable-next-line @jessie.js/no-nested-await
+      const tmpSnap = await fs.open(snapPath, 'w');
+      // eslint-disable-next-line @jessie.js/no-nested-await
+      await tmpSnap.writeFile(
+        // @ts-expect-error incorrect typings, does support AsyncIterable
+        snapshotStream,
+      );
+      // eslint-disable-next-line @jessie.js/no-nested-await
+      await tmpSnap.close();
+    } catch (e) {
+      // eslint-disable-next-line @jessie.js/no-nested-await
+      await cleanup();
+      throw e;
+    }
+
+    return harden({
+      snapPath,
+      afterSpawn,
+      cleanup,
+    });
+  };
+
+  const makeLoadSnapshotHandlerWithPipe = async () => {
+    let done = Promise.resolve();
+
+    const cleanup = async () => done;
+
+    /** @param {Writable} loadSnapshotsStream */
+    const afterSpawn = async loadSnapshotsStream => {
+      assert(snapshotStream);
+      const destStream = loadSnapshotsStream;
+
+      const sourceStream = Readable.from(snapshotStream);
+      sourceStream.pipe(destStream, { end: false });
+
+      done = finished(sourceStream);
+      done.catch(noop).then(() => sourceStream.unpipe(destStream));
+    };
+
+    return harden({
+      snapPath: `@8:${safeHintFromDescription(snapshotDescription)}`,
```

Comment:
that `8` is a bit magic, maybe define `SNAPSHOT_LOAD_FD = 8` up above, use it here, and mention it below in the big list of numbered FDs, in case we think to add one in the middle in the future, for some reason